### PR TITLE
Integrate cpp package

### PR DESCRIPTION
**Pull request by lx75249**
As mentioned in dmlc/mxnet.cpp#71, we propose to merge mxnet.cpp into mxnet project.

It will attract more attention to mxnet.cpp, and it will also enable integrated test with mxnet, so that future checkins to mxnet will not break mxnet.cpp. Meanwhile, it is important for industrial environment (e.g, data center), where python and other dependencies are normally not installed.

### CODE REVIEW

**Comment by piiswrong** [comment]
I think merging is a good idea. Not sure about submodule though. We have too many submodules already

**Comment by lx75249** [comment]
I also prefer copying mxnet.cpp directly to using submodule, as it allows direct reference to dependencies like c_api.h and dmlc-core. However, this will diverge the cpp package into two versions, one in dmlc/mxnet.cpp, and the other one in dmlc/mxnet, which is confusing. Maybe we can just move all the development to mxnet?

**Comment by piiswrong** [comment]
Yes. We can mark mxnet.cpp as discontinued.

**Comment by hjk41** [comment]
We will mark dmlc/mxnet.cpp as deprecated. Future checkins will be made only in dmlc/mxnet.

**Comment by piiswrong** [comment]
Can one of the admins verify this patch?

**Comment by piiswrong** [comment]
any updates on this?

**Comment by lx75249** [comment]
I'm updating it to remove the submodule, and to include headers from mxnet.

**Comment by lx75249** [comment]
I have some question about BatchNorm. It recently added two arguments gamma and beta, which don't have default values. But I found that in python example codes, these arguments can be omitted. This makes mxnet.cpp unable to generate correct function signature for BatchNorm.

**Comment by piiswrong** [comment]
not sure what you mean. Are you talking about initializer? what default value?

**Comment by lx75249** [comment]
For example, the default value for eps in batchnorm is 1e-3 (https://github.com/dmlc/mxnet/blob/master/src/operator/batch_norm-inl.h#L37), and gamma and beta have no default values(https://github.com/dmlc/mxnet/blob/master/src/operator/batch_norm.cc#L81), while some example code calls batchnorm without specifying gamma and beta (https://github.com/dmlc/mxnet/blob/master/tests/python/common/models.py#L16).

**Comment by piiswrong** [comment]
gamma and beta are weights, not parameters. their default values are set by initializer

**Comment by piiswrong** [comment]
ndarray-or-symbol replaced NDArray and Symbol for types.

**Comment by lx75249** [comment]
This pull request is ready for review now.

**Comment by piiswrong** [comment]
Any idea why windows tests failed?

**Comment by lx75249** [comment]
Fixed
